Change the method of desync detection in RageSoundReader_Merge

sukibaby opened the pull request:
## RageSoundReader_Merge::Read()

The first major change here is in the loop responsible for reading audio frames from multiple sources. It is written in a way that the compiler is unlikely to optimize for the case where the threshold is not exceeded. I've changed it to only execute the resynchronization logic when the threshold is exceeded (the existing code awkwardly checks if the threshold has _not been_ exceeded, and `continue`s if so). 

The second major change is to directly initialize the `aNextSourceFrames` and `aRatios` vectors with the size of the vector which holds pointers to `RageSoundReader` objects. We then remove the `aNextSourceFrames.front()` and `aRatios.front()` assignments from the single-source case, since the information stored in these variables only lasts during the lifetime of the function. For the single-source case, the function will complete without any of these variables being referenced again. It is perfectly safe to remove these assignments.

There are some small things here, like adding missing braces, or `std::` prefixes to certain function calls.

teejusb commented on src/RageSoundReader_Merge.cpp:
```diff
@@ -227,18 +223,16 @@ int RageSoundReader_Merge::Read( float *pBuffer, int iFrames )
 		iFrames = pSound->Read( pBuffer, iFrames );
 		if( iFrames > 0 )
-			m_iNextSourceFrame += static_cast<int>((iFrames * m_fCurrentStreamToSourceRatio) + 0.5 );
-		aNextSourceFrames.front() = pSound->GetNextSourceFrame();
```
Why are we removing these two lines?

sukibaby replied:
Please see the longer comment below

teejusb commented:
The commit messages don't say anything reasonable. There's nothing particularly being "fixed" and there's no description of what you're actually doing here.

sukibaby commented:
> The commit messages don't say anything reasonable. There's nothing particularly being "fixed" and there's no description of what you're actually doing here.

I will demonstrate a general overview of the logic flow of RageSoundReader_Merge::Read below along with a flow chart. After the flow chart I will describe the change in greater detail. I will also squash the commits in order to re-write the commit message.

This function is responsible for **detecting desync and correcting it**. A threshold which is "number of frames away from the expected position" we consider tolerable is set to 16 (this is the value it's always been). This is checked inside a loop. We are checking if we are *less or equal to the threshold* and continuing the loop if so. Because of this, it allows differences up to the threshold limit to go uncorrected. This allows **small desyncs to accumulate over time** because they will be considered "good enough" until an event exceeding the threshold occurs.  This PR changes the logic flow to **immediately correct any desync** before continuing the loop.

RageSoundReader_Merge handles merging multiple sound sources into a single stream. It is incredibly important. This is where desync is identified, tracked and resolved. All audio goes thru this. It can set the playback position, read audio data, resample when needed, etc.

Below,  the "minimum position" refers to the earliest frame position among all sound sources being merged. Determining if any sound sources are lagging behind what is expected is how desync is identified and tracked.  

Below,  "mix buffer" refers to RageSoundMixBuffer.

```mermaid
graph TD
    A[RageSoundReader_Merge::Read] --> B{Are there any sounds?}
    B -- No --> C[Return EOF]
    B -- Yes --> D[Initialize variables & fill with sound data]
    D --> E[Find the earliest sound]
    E --> F[Delay any sounds that are early so that they line up later]
    F --> G{Do we need to adjust any sound positions?}
    G -- Yes --> H[Adjust the position and return 0]
    G -- No --> J[Find the minimum position]
    J --> K{Is the difference in positions equal to or less than the acceptable threshold?}
    K -- No --> L[Clamp iFrames]
    L --> M
    K -- Yes --> M{Is there only one sound playing?}
    M -- Yes --> P[Write directly into buffer, update multi-source sync info & return]

    M -- No --> Q[Initialize mix buffer & read each sound to play]
    Q --> S[Check for resync and read frames]
    S --> T[Write to mix buffer]
    T --> U[Read mixed frames into pBuffer]
    U --> V[Update multi-source sync info & return]
```

Notice something is funny with the logic flow of this function.  We are checking if we are *less or equal to the threshold* and continuing the loop if so. Because of this, it allows differences up to the threshold limit to go uncorrected. This allows **small desyncs to accumulate over time** because they will be considered "good enough" until an event exceeding the threshold occurs.  

I fix this by changing the logic from `<=` plus `continue` (less effective) to `>` controlling an if loop (more effective).  It is not just about the specific threshold value -  the method of detection directly impacts the detection and handling of desync. With `<=`, the loop continues without taking any corrective action.  By changing it to immediately correct as soon as anything greater than the threshold is detected, we get significantly improved in-game sync. There is no possibility of accumulating desyncs with this PR. Combined with the preceding sync fixes, this gives us basically perfect sync stability. In testing on both WaveOut and DirectSound I had several of the best timing plays of my life with up to 3 hour in-game time in each instance.



Here is a comparison of the relevant section
new
```
	for (size_t i = 0; i < m_aSounds.size(); ++i)
	{
		if (Difference(aNextSourceFrames[i], iMinPosition) > ERROR_CORRECTION_THRESHOLD)
		{
			/* A sound is being delayed to resync it; clamp the number of frames we
			 * read now, so we don't advance past it. */
			int iMaxSourceFramesToRead = aNextSourceFrames[i] - iMinPosition;
			int iMaxStreamFramesToRead = static_cast<int>((iMaxSourceFramesToRead / m_fCurrentStreamToSourceRatio) + 0.5 );
			iFrames = std::min( iFrames, iMaxStreamFramesToRead );
//			LOG->Warn( "RageSoundReader_Merge: sound positions moving at different rates" );
		}
	}
```

vs. old
```
		for( unsigned i = 0; i < m_aSounds.size(); ++i )
		{
			if( Difference(aNextSourceFrames[i], iMinPosition) <= ERROR_CORRECTION_THRESHOLD )
				continue;

			/* A sound is being delayed to resync it; clamp the number of frames we
			 * read now, so we don't advance past it. */
			int iMaxSourceFramesToRead = aNextSourceFrames[i] - iMinPosition;
			int iMaxStreamFramesToRead = static_cast<int>((iMaxSourceFramesToRead / m_fCurrentStreamToSourceRatio) + 0.5 );
			iFrames = std::min( iFrames, iMaxStreamFramesToRead );
//			LOG->Warn( "RageSoundReader_Merge: sound positions moving at different rates" );
		}
```

I have removed the updating of the next frame position (removed lines 231 & 232) from **only the single-source case** since it is never needed, and never used.  The single-source case is commonly used; it's what is used when in-game and only a song is playing, without assist tick. The updated frame position is not needed, and not used - the vectors will be re-initialized and re-filled at the beginning of the next `Read` call. Additionally, this information is only used for multi-source mixing.  You can see this at the end of the "Is there only one sound playing? -> Yes" section of the flowchart. There is no harm done when we start using a multi-source source afterwards; it will be in the results screen or song wheel and the song will be done at that point.

sukibaby committed:
RageSoundReader_Merge fixes

This is the fixed Merge from the 2025-1-26 test build. It will more strictly check the drift with each iteration of the m_aSounds loop, whereas the original implementation can allow small amounts of desync to accumulate over time if none of them exceed the threshold.

teejusb commented:
This PR is essentially no-op but it's fine to merge since it does some things like using static_cast which is preferred
